lib/Dialect/RTG/Transforms/ElaborationPass.cpp

```diff
@@ -0,0 +1,568 @@
+    for (auto *el : val->getSet()) {
+      auto materialized = materialize(el, builder.getBlock(), loc);
+      if (!materialized)
+        return Value();
+
+      elements.push_back(materialized);
+    }
+
+    auto res = builder.create<SetCreateOp>(loc, val->getType(), elements);
+    materializedValues[{val, builder.getBlock()}] = res;
+    return res;
+  }
+
+private:
+  /// Cache values we have already materialized to reuse them later. We start
+  /// with an insertion point at the start of the block and cache the (updated)
+  /// insertion point such that future materializations can also reuse previous
+  /// materializations without running into dominance issues (or requiring
+  /// additional checks to avoid them).
+  DenseMap<std::pair<ElaboratorValue *, Block *>, Value> materializedValues;
+
+  /// Cache the builders to continue insertions at their current insertion point
+  /// for the reason stated above.
+  DenseMap<Block *, OpBuilder> builderPerBlock;
+};
+
+/// Used to signal to the elaboration driver whether the operation should be
+/// removed.
+enum class DeletionKind { Keep, Delete };
+
+/// Interprets the IR to perform and lower the represented randomizations.
+class Elaborator : public RTGOpVisitor<Elaborator, FailureOr<DeletionKind>,
+                                       function_ref<void(Operation *)>> {
+public:
+  using RTGBase = RTGOpVisitor<Elaborator, FailureOr<DeletionKind>,
+                               function_ref<void(Operation *)>>;
+  using RTGBase::visitOp;
+
+  Elaborator(SymbolTable &table, std::mt19937 &rng) : rng(rng) {}
+
+  /// Helper to perform internalization and keep track of interpreted value for
+  /// the given SSA value.
+  template <typename ValueTy, typename... Args>
+  void internalizeResult(Value val, Args &&...args) {
+    // TODO: this isn't the most efficient way to internalize
+    auto ptr = std::make_unique<ValueTy>(std::forward<Args>(args)...);
+    auto *e = ptr.get();
+    auto [iter, _] = interned.insert({e, std::move(ptr)});
+    state[val] = iter->second.get();
+  }
+
+  /// Print a nice error message for operations we don't support yet.
+  FailureOr<DeletionKind>
+  visitUnhandledOp(Operation *op,
+                   function_ref<void(Operation *)> addToWorklist) {
+    return op->emitError("elaboration not supported");
+  }
+
+  FailureOr<DeletionKind>
+  visitExternalOp(Operation *op,
+                  function_ref<void(Operation *)> addToWorklist) {
+    // TODO: we only have this to be able to write tests for this pass without
+    // having to add support for more operations for now, so it should be
+    // removed once it is not necessary anymore for writing tests
+    if (op->use_empty()) {
+      for (auto &operand : op->getOpOperands()) {
+        Value val = materializer.materialize(state.at(operand.get()),
+                                             op->getBlock(), op->getLoc());
+        if (!val)
+          return op->emitError("failed to materialize value");
```

Comment:
I'd probably be a good idea to pass some diagnostic to the materializer and print more specific errors.